docs/RFCS/20210526_expression_indexes.md

```diff
@@ -0,0 +1,288 @@
+
+1. It has a name with the prefix `crdb_idx_expr`.
+2. It is a hidden virtual column.
+3. It has a matching expression.
+
+If no such column exists, a new hidden virtual column is created.
+
+### Dropping an expression index
+
+When an expression index is dropped, the hidden virtual column(s) that represent
+its expression(s) are dropped, if possible.
+
+Notice that an index descriptor includes no information to denote whether or not
+it is an expression index. An expression index loses this identity as soon as
+the AST representing it is converted into an index descriptor. This poses two
+challenges:
+
+1. How can we drop virtual columns used by an expression index if we cannot
+determine an index is expression-based.
+2. How can we differentiate an indexed virtual column created for an expression
+index with one created manually by a user. We want to drop only the former, not
+the latter.
+
+To address these challenges, when dropping any index, an indexed column is
+dropped if:
+
+1. It has a name with the prefix `crdb_idx_expr`.
+2. It is a hidden virtual column.
+3. It is not referenced by another index, partial index predicate, constraint,
+or computed column expression.
+
+### Displaying an expression index
+
+When displaying an expression index to a user, the expressions it indexes are
+displayed, not the names of virtual columns that represent the expression.
+
+As mentioned in the section above, nothing in an index descriptor denotes that
+it is an expression index. This creates challenges in displaying expression
+indexes similar to the challenges in dropping them.
+
+To address these challenges, when displaying any index, an indexed virtual
+column's expression is displayed rather than its name if:
+
+1. It has a name with the prefix `crdb_idx_expr`.
+2. It is a hidden virtual column.
+
+When the table containing an expression index is displayed, the hidden virtual
+column(s) that the index relies on is displayed as `NOT VISIBLE`. This is
+standard for all hidden columns.
+
+The `crdb_idx_expr` column name prefix communicates that a column originated
+from an expression index. This is critical for satisfying **C**. It must be
+communicated that a column was created for an expression index to avoid creating
+a duplicate column when the output of `SHOW CREATE TABLE` is executed. The
+output of `SHOW CREATE TABLE` cannot communicate arbitrary information about
+columns; it must be valid SQL.
+
+## Putting it all together
+
+The following example shows the outcome of creating, dropping, and displaying
+expression indexes.
+
+```sql
+CREATE TABLE t (
+    k INT PRIMARY KEY,
+    a INT,
+    b INT,
+    INDEX t_a_plus_b_idx ((a + b))
+);
+
+SELECT create_statement FROM [SHOW CREATE TABLE t];
+                  create_statement
+------------------------------------------------------
+CREATE TABLE public.t (
+    k INT8 NOT NULL,
+    a INT8 NULL,
+    b INT8 NULL,
+    crdb_idx_expr INT8 NOT VISIBLE NULL AS (a + b) VIRTUAL,
```

Comment:
Can we set out to omit this row? I think that'd be nice. What's the benefit of displaying it?

Reply:
Ha... I guess I was so wrapped up in writing this up that I didn't realize we can omit this based on the prefix. Let me ruminate about this for a bit before committing to it.

Reply:
Yes or omit it based on the fact that the column is expression based as noted by some other state in the descriptor (my current inclination).

Reply:
I mention some options like this in Alternatives, but they don't seem great.

Reply:
I'm skeptical of the cons of doing the bookkeeping in the descriptor explicitly. It doesn't seem like very much complexity.